src/client.rs

```diff
@@ -796,9 +802,18 @@ impl State {
         client
             .engine
             .crypto_context_mut()
-            .derive_keys(cipher_suite, &client_random_buf, &server_random_buf)
+            .derive_keys(
+                cipher_suite,
+                &client_random_buf,
+                &server_random_buf,
+                &mut out,
+                &mut scratch,
+            )
             .map_err(|e| Error::CryptoError(format!("Failed to derive keys: {}", e)))?;
```

Comment:
should you push back buffers on fail here?

Reply:
It doesn't matter hugely. Buf will still be zeroed (at least when we reintroduce zeroise). I think it's enough if the happy paths return to buf.